### PR TITLE
Bug 1120878 - [Contacts] View vcard activity should disable edit butt…

### PR DESCRIPTION
**Pull request by adrm**
…on if the vcard is an attachment that is already on the device

### CODE REVIEW

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=bef553dfc7e9)

File: apps/communications/contacts/js/activities.js [hunk 1]
```diff
@@ -95,11 +95,10 @@ var ActivityHandler = {
       case 'open':
         if (this.isvCardActivity(activity)) {
           var self = this;
-          var dependency =
-                         '/shared/js/contacts/import/utilities/vcard_reader.js';
-          LazyLoader.load(dependency, function() {
-            self.getvCardReader(activity.source.data.blob,
-                                           self.readvCard.bind(self, activity));
+          var deps = '/shared/js/contacts/import/utilities/vcard_reader.js';
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/activities.js]: is this change needed? if it is not related to this PR, please drop it

File: apps/communications/contacts/js/views/details.js [hunk 2]
```diff
@@ -214,6 +214,17 @@ contacts.Details = (function() {
     // Initially enabled and only disabled if necessary
     editContactButton.removeAttribute('disabled');
 
+    socialTemplate.style.display = 'inherit';
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/details.js]: no. if we need something like that it should be done through CSS

File: apps/communications/contacts/js/contacts.js [hunk 3]
```diff
@@ -88,26 +88,34 @@ var Contacts = (function() {
...
         initContactsList();
         initDetails(function onInitDetails() {
-          if (params == -1 || !('id' in params)) {
-            console.error('Param missing');
-            return;
-          }
-          var id = params.id;
-          cList.getContactById(id, function onSuccess(savedContact) {
-            currentContact = savedContact;
+          if (params != -1) {
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/contacts.js]: we would need to document the different code paths and possibilities here

File: apps/communications/contacts/js/views/details.js [hunk 4]
```diff
@@ -200,7 +200,7 @@ contacts.Details = (function() {
     cover.addEventListener('touchstart', onTouchStart, true);
   };
 
-  var render = function cd_render(currentContact, fbContactData) {
+  var render = function cd_render(currentContact, fbContactData, staticMode) {
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/details.js]: please document what does it mean 'staticMode' . What does it mean? can we have another more meaningful name?

File: apps/communications/contacts/js/views/details.js [hunk 5]
```diff
@@ -214,6 +214,17 @@ contacts.Details = (function() {
     // Initially enabled and only disabled if necessary
     editContactButton.removeAttribute('disabled');
 
+    socialTemplate.style.display = 'inherit';
+    duplicateTemplate.style.display = 'inherit';
+    favoriteMessage.style.display = 'flex';
+
+    if (staticMode) {
+      editContactButton.setAttribute('disabled', 'disabled');
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/details.js]: CSS!

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=49d93b9058a3)

File: apps/communications/contacts/js/contacts.js [hunk 6]
```diff
@@ -88,26 +88,38 @@ var Contacts = (function() {
...
-          if (params == -1 || !('id' in params)) {
-            console.error('Param missing');
-            return;
-          }
-          var id = params.id;
-          cList.getContactById(id, function onSuccess(savedContact) {
-            currentContact = savedContact;
+          // At this point, a parameter is required. If the parameter is an id,
+          // the corresponding contact is loaded from the device. If the
+          // parameter tells us that a contact blob is attached to the activity,
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/contacts.js]: please clarify the comments

File: apps/communications/contacts/js/contacts.js [hunk 7]
```diff
@@ -88,26 +88,38 @@ var Contacts = (function() {
...
-            showApp();
-          }, function onError() {
-            console.error('Error retrieving contact');
-          });
+                navigation.go(sectionId, 'right-left');
+                showApp();
+              }, function onError() {
+                console.error('Error retrieving contact');
+              });
+            } else if (params.mozContactParam) {
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/contacts.js]: comment here what contains mozContactParam

File: apps/communications/contacts/js/contacts.js [hunk 8]
```diff
@@ -88,26 +88,38 @@ var Contacts = (function() {
...
+              var contact = ActivityHandler.mozContactParam;
+              contactsDetails.render(contact, null, true);
+              navigation.go(sectionId, 'right-left');
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/contacts.js]: 'activity-popup' should be the transition used in this case

File: apps/communications/contacts/js/views/details.js [hunk 9]
```diff
@@ -214,6 +216,11 @@ contacts.Details = (function() {
     // Initially enabled and only disabled if necessary
     editContactButton.removeAttribute('disabled');
 
+    if (readOnly) {
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/details.js]: reset the state to the contrary in case it is not readonly

File: apps/communications/contacts/js/views/details.js [hunk 10]
```diff
@@ -200,7 +202,7 @@ contacts.Details = (function() {
     cover.addEventListener('touchstart', onTouchStart, true);
   };
 
-  var render = function cd_render(currentContact, fbContactData) {
+  var render = function cd_render(currentContact, fbContactData, readOnly) {
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/details.js]: add a comment indicating what is readonly and in what case is being used

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=e5221a26fdbb)

File: apps/communications/contacts/js/views/details.js [hunk 11]
```diff
@@ -213,6 +216,14 @@ contacts.Details = (function() {
 
     // Initially enabled and only disabled if necessary
     editContactButton.removeAttribute('disabled');
+    editContactButton.classList.remove('hide');
+
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/details.js]: nit: please group these two sentences

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=635407667643)

File: apps/communications/contacts/js/contacts.js [hunk 12]
```diff
@@ -88,26 +88,40 @@ var Contacts = (function() {
...
-          if (params == -1 || !('id' in params)) {
-            console.error('Param missing');
-            return;
-          }
-          var id = params.id;
-          cList.getContactById(id, function onSuccess(savedContact) {
-            currentContact = savedContact;
+          // At this point, a parameter is required. If the parameter is an id,
+          // the corresponding contact is loaded from the device.
+          if (params != -1) {
```
Comment (arcturus) [inline apps/communications/contacts/js/contacts.js]: Too much indentation, we could early exit avoiding the last `else`

**Comment by arcturus** [comment]
Closing and reopening to trigger again the tests

**Comment by try-server-hook** [comment]
[arcturus Francisco Jordano (arcturus)](https://github.com/arcturus) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=cf1b5a6ca1f9)

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=39cc537fae7b)

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=8ae756c8b603)

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=b8ae19a8369e)

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=eb0b9051fb65)

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=c1910d953bd3)

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=8d0abce9cc31)

File: apps/communications/contacts/js/activities.js [hunk 13]
```diff
@@ -125,8 +125,13 @@ var ActivityHandler = {
 
   renderOneContact: function(contact, activity) {
     this.mozContactParam = contact;
-    activity.source.data.params = {'mozContactParam': true};
-    this.launch_activity(activity, 'view-contact-form');
+    var data = activity.source.data;
+    data.params = {'mozContactParam': true};
+    if (typeof data.allowSave === 'undefined' || data.allowSave) {
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/activities.js]: why don't we just check for data.allowSave === true ?

File: apps/communications/contacts/test/marionette/activities_test.js [hunk 14]
```diff
@@ -55,21 +46,98 @@ marionette('Contacts > Activities', function() {
...
+            allowSave: false
+          }
+        });
+      }, [fs.readFileSync(__dirname + '/data/vcard_4.vcf', 'utf8')]);
+
+      client.switchToFrame();
+      client.apps.switchToApp(Contacts.URL, 'contacts');
+      client.helper.waitForElement(selectors.details);
+
+      assertDetailsData();
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/test/marionette/activities_test.js]: we would need to check that the buttons we don't want to appear actually do not appear

File: apps/communications/contacts/test/marionette/activities_test.js [hunk 15]
```diff
@@ -55,21 +46,98 @@ marionette('Contacts > Activities', function() {
...
+            allowSave: true
+          }
+        });
+      }, [fs.readFileSync(__dirname + '/data/vcard_4.vcf', 'utf8')]);
+
+      client.switchToFrame();
+      client.apps.switchToApp(Contacts.URL, 'contacts');
+      client.helper.waitForElement(selectors.form);
+
+      assertFormData();
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/test/marionette/activities_test.js]: shouldn't be test here that the 'done' button is enabled?

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=70ce896370ca)

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=52cff71ee6dd)

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=cb41a3eb460c)

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=b476a0bef3c0)

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=ff1cae0b2880)
